src/security/tests/acl_api_tests.c

```diff
@@ -1342,6 +1344,654 @@ test_acl_remove_ace_not_found(void **state)
+
+	ace[0] = daos_ace_create(DAOS_ACL_EVERYONE, NULL);
+	ace[1] = daos_ace_create(DAOS_ACL_USER, "user1@");
+	ace[2] = daos_ace_create(DAOS_ACL_EVERYONE, NULL);
+	acl = daos_acl_create(ace, num_aces);
+
+	assert_int_equal(daos_acl_validate(acl), -DER_INVAL);
+
+	daos_acl_free(acl);
+	free_all_aces(ace, num_aces);
+}
+
+static void
+test_acl_is_valid_duplicate_user(void **state)
+{
+	struct daos_acl *acl;
+	size_t		num_aces = 3;
+	struct daos_ace *ace[num_aces];
+
+	ace[0] = daos_ace_create(DAOS_ACL_USER, "user1@");
+	ace[1] = daos_ace_create(DAOS_ACL_USER, "anotheruser@");
+	ace[2] = daos_ace_create(DAOS_ACL_USER, "user1@");
+	acl = daos_acl_create(ace, num_aces);
+
+	assert_int_equal(daos_acl_validate(acl), -DER_INVAL);
+
+	daos_acl_free(acl);
+	free_all_aces(ace, num_aces);
+}
+
+static void
+test_acl_is_valid_duplicate_group(void **state)
+{
+	struct daos_acl *acl;
+	size_t		num_aces = 3;
+	struct daos_ace *ace[num_aces];
+
+	ace[0] = daos_ace_create(DAOS_ACL_GROUP, "grp1@");
+	ace[1] = daos_ace_create(DAOS_ACL_GROUP, "anothergroup@");
+	ace[2] = daos_ace_create(DAOS_ACL_GROUP, "grp1@");
+	acl = daos_acl_create(ace, num_aces);
+
+	assert_int_equal(daos_acl_validate(acl), -DER_INVAL);
+
+	daos_acl_free(acl);
+	free_all_aces(ace, num_aces);
+}
+
+static struct daos_acl *
+acl_create_in_exact_order(struct daos_ace *ace[], size_t num_aces)
+{
+	struct daos_acl	*acl;
+	uint8_t		*pen;
+	size_t		i;
+
+	acl = daos_acl_create(ace, num_aces);
+
+	/* Create probably reordered our input - rewrite in exact order */
+	pen = acl->dal_ace;
+	for (i = 0; i < num_aces; i++) {
+		ssize_t ace_len = daos_ace_get_size(ace[i]);
+
+		assert_true(ace_len > 0);
+
+		memcpy(pen, ace[i], ace_len);
+		pen += ace_len;
+	}
+
+	return acl;
+}
+
+static bool
+needs_name(enum daos_acl_principal_type type)
+{
+	return (type == DAOS_ACL_USER || type == DAOS_ACL_GROUP);
+}
+
+static void
+expect_acl_invalid_bad_ordering(enum daos_acl_principal_type type1,
+		enum daos_acl_principal_type type2)
+{
+	struct daos_acl *acl;
+	size_t		num_aces = 2;
+	struct daos_ace *ace[num_aces];
+	const char	*name1 = NULL;
+	const char	*name2 = NULL;
+
+	if (needs_name(type1)) {
+		name1 = "name1@";
+	}
+
+	if (needs_name(type2)) {
+		name2 = "name2@";
+	}
+
+	ace[0] = daos_ace_create(type1, name1);
+	ace[1] = daos_ace_create(type2, name2);
+	acl = acl_create_in_exact_order(ace, num_aces);
+
+	assert_int_equal(daos_acl_validate(acl), -DER_INVAL);
+
+	daos_acl_free(acl);
+	free_all_aces(ace, num_aces);
+}
+
+static void
+test_acl_is_valid_bad_ordering(void **state)
+{
+	expect_acl_invalid_bad_ordering(DAOS_ACL_USER, DAOS_ACL_OWNER);
+	expect_acl_invalid_bad_ordering(DAOS_ACL_OWNER_GROUP, DAOS_ACL_USER);
+	expect_acl_invalid_bad_ordering(DAOS_ACL_GROUP, DAOS_ACL_OWNER_GROUP);
+	expect_acl_invalid_bad_ordering(DAOS_ACL_EVERYONE, DAOS_ACL_GROUP);
+	expect_acl_invalid_bad_ordering(DAOS_ACL_EVERYONE, DAOS_ACL_OWNER);
+}
+
+
+static void
+test_acl_random_buffer(void **state)
```

Comment:
Have you tried to run this unit test in a loop to make sure that it never crashes?

Reply:
In theory it could crash if the top level header looks valid, has a length set longer than the buffer, and the ACEs also appear to be valid at least to the end of the buffer. Unlikely but possible. I'll tighten this up.